numpy/core/src/npysort/qsort-32bit-avx512.h.src

```diff
@@ -0,0 +1,518 @@
+            left += 16;
+        }
+        // partition the current vector and save it on both sides of the array
+        npy_int amount_gt_pivot = partition_vec_@vsuffix@(arr, l_store, r_store + 16, curr_vec, pivot_vec, &min_vec, &max_vec);;
+        r_store -= amount_gt_pivot; l_store += (16 - amount_gt_pivot);
+    }
+
+    /* partition and save vec_left and vec_right */
+    npy_int amount_gt_pivot = partition_vec_@vsuffix@(arr, l_store, r_store + 16, vec_left, pivot_vec, &min_vec, &max_vec);
+    l_store += (16 - amount_gt_pivot);
+    amount_gt_pivot = partition_vec_@vsuffix@(arr, l_store, l_store + 16, vec_right, pivot_vec, &min_vec, &max_vec);
+    l_store += (16 - amount_gt_pivot);
+    *smallest = _mm512_reduce_min_@vsuffix@(min_vec);
+    *biggest = _mm512_reduce_max_@vsuffix@(max_vec);
+    return l_store;
+}
+
+static NPY_INLINE NPY_GCC_OPT_3 NPY_GCC_TARGET_AVX512_SKX
+void qsort_@type@(@type_t@* arr, npy_intp left, npy_intp right, npy_int max_iters)
+{
+    if (max_iters <= 0) {
+        heapsort_@type@((void*)(arr + left), right + 1 - left, NULL);
+        return;
+    }
+    if (right + 1 - left <= 128) {
+        sort_128_@vsuffix@(arr + left, right + 1 -left);
+        return;
+    }
+
+    @type_t@ pivot = get_pivot_@vsuffix@(arr, left, right);
+    @type_t@ smallest = @TYPE_MAX_VAL@;
+    @type_t@ biggest = @TYPE_MIN_VAL@;
+    npy_intp pivot_index = partition_avx512_@vsuffix@(arr, left, right+1, pivot, &smallest, &biggest);
+    if (pivot != smallest)
+        qsort_@type@(arr, left, pivot_index - 1, max_iters - 1);
+    if (pivot != biggest)
+        qsort_@type@(arr, pivot_index, right, max_iters - 1);
+}
+/**end repeat**/
+
+static NPY_INLINE NPY_GCC_OPT_3 NPY_GCC_TARGET_AVX512_SKX
+npy_intp replace_nan_with_inf(npy_float* arr, npy_intp arrsize)
+{
+    npy_intp nan_count = 0;
+    __mmask16 loadmask = 0xFFFF;
+    while (arrsize > 0) {
+        if (arrsize < 16) {
+            loadmask = (0x0001 << arrsize) - 0x0001;
+        }
+        __m512 in_zmm = _mm512_maskz_loadu_ps(loadmask, arr);
+        __mmask16 nanmask = _mm512_cmpunord_ps_mask(in_zmm, in_zmm);
+        nan_count += _mm_popcnt_u32((npy_int) nanmask);
+        _mm512_mask_storeu_ps(arr, nanmask, ZMM_MAX_FLOAT);
+        arr += 16;
+        arrsize -= 16;
+    }
+    return nan_count;
+}
+
+static NPY_INLINE NPY_GCC_OPT_3 NPY_GCC_TARGET_AVX512_SKX
+void replace_inf_with_nan(npy_float* arr, npy_intp arrsize, npy_intp nan_count)
+{
+    for (npy_intp ii = arrsize-1; nan_count > 0; --ii) {
+        arr[ii] = NPY_NANF;
+        nan_count -= 1;
+    }
+}
+
+/**begin repeat
+ *
+ * #type = int, uint, float#
+ * #type_t = npy_int, npy_uint, npy_float#
+ * #FIXNAN = 0, 0, 1#
+ */
+
+void avx512_qsort_@type@(@type_t@* arr, npy_intp arrsize)
+{
+    if (arrsize > 1) {
+#if @FIXNAN@
+        npy_intp nan_count = replace_nan_with_inf(arr, arrsize);
+#endif
+        qsort_@type@(arr, 0, arrsize-1, 2*log2(arrsize));
```

Comment:
Maybe replace `log2` with `(63 - __builtin_clzll(arrsize))` (`lzcntq`). AFAIK all machines that support `avx512` have fast `__builtin_clzll` and the precision from `log2` isn't really needed.